contracts/implementation/InvariantCheck.sol

```diff
@@ -0,0 +1,66 @@
+//SPDX-License-Identifier: CC-BY-NC-ND-4.0
+pragma solidity 0.8.7;
+
+import "../interfaces/IInvariantCheck.sol";
+import "../interfaces/IPoolCommitter.sol";
+import "../interfaces/IPausable.sol";
+import "../interfaces/ILeveragedPool.sol";
+import "../interfaces/IPoolFactory.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/// @title The contract for checking invariants and pausing if necessary
+/// @notice Every time certain functions are called, known invariants are checked and if any do not hold, contracts are paused.
+contract InvariantCheck is IInvariantCheck {
+    IPoolFactory public immutable poolFactory;
+
+    constructor(address _factory) {
```

Comment:
Whilst I have a PR open for it currently (see #205), we might as well add NatSpec comments now as a general practice.